### PR TITLE
Trainer support for iterabledataset

### PR DESCRIPTION
**Pull request by Pradhy729**
To fix --> #5829

### CODE REVIEW

File: src/transformers/trainer.py
Comment (sshleifer) [inline src/transformers/trainer.py:252]: why about TPU + Iterable Dataset case?
I think the train_sampler logic needs to be unnested.
(Full disclosure I don't know this code well.)

**Review by sshleifer — COMMENTED** [review]
would be great to add a test that was breaking and now works.

File: src/transformers/trainer.py
Comment (Pradhy729) [reply src/transformers/trainer.py:252]: Yeah I agree this looks ugly. I'll update it - thanks.

**Comment by Pradhy729** [comment]
I will add a test for it.

**Comment by Pradhy729** [comment]
@sshleifer --> I have made the changes you recommended and added a test that should fail without this change. Please review and let me know if there is anything else needed.

File: tests/test_trainer.py
Comment (sshleifer) [inline tests/test_trainer.py:193]: Please use the appropriate "tiny" model from  https://huggingface.co/models?search=tiny. The test should run in < 1s. You can time it by passing `--durations 0` to pytest.

File: tests/test_trainer.py
Comment (sshleifer) [inline tests/test_trainer.py:172]: you could change this too.

File: src/transformers/trainer.py
Comment (sshleifer) [inline src/transformers/trainer.py:236]: why is this nested?

File: src/transformers/trainer.py
Comment (sshleifer) [inline src/transformers/trainer.py:238]: use elif, no need for nesting.

**Review by sshleifer — COMMENTED** [review]
Thanks for iterating!

**Comment by Pradhy729** [comment]
OK Will make the changes. Also do you know where to place the dataset class to prevent failure in the tf tests?

**Comment by Pradhy729** [comment]
I've made the changes. Please review and feel free to edit as needed and merge

File: tests/test_trainer.py
Comment (Pradhy729) [reply tests/test_trainer.py:172]: Which is the tiny version for mrpc task?

Comment (sshleifer) [reply tests/test_trainer.py:172]: just leave, it. Not fair of me to ask you to get to the bottom and they check the loss below. Sorry.

File: tests/test_trainer.py
Comment (sshleifer) [inline tests/test_trainer.py:160]: probably doesn't need to be nested either afaict.

**Review by sshleifer — APPROVED** [review]
lgtm, we will wait for @julien-c to merge

File: tests/test_trainer.py
Comment (sshleifer) [reply tests/test_trainer.py:160]: do might have to define it inside your test method. I think attributes declared above `__init__` are not protected by `@require_torch` :(

Comment (sshleifer) [reply tests/test_trainer.py:160]: or add an `if is_torch_available` outside the test class and define inside.

**Comment by Pradhy729** [comment]
Should be good to go. I removed the WIP. Thanks.

**Comment by julien-c** [comment]
LGTM but will wait for review from @sgugger too

**Review by sgugger — APPROVED** [review]
LGTM, thanks for fixing this!

**Pull request merged**
